File: go/39.combination-sum.go

var result [][]int

func combinationSum(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	result = [][]int{}
	findCombination([]int{}, candidates, target)

	return result
}

func findCombination(combination []int, candidates []int, target int) {
	if target == 0 {
		result = append(result, makeSliceCopy(combination))
	} else if target < 0 {
		return 
	} else {
		for i := 0; i < len(candidates); i++ {
			findCombination(append(combination, candidates[i]), candidates[i:], target - candidates[i])
		}
	}
}

func makeSliceCopy(s []int) []int {
	copy := make([]int, len(s))
	for i, e := range s {
		copy[i] = e
	}
	return copy
}

File: csharp/39.combination-sum.cs
/*
 * @lc app=leetcode id=39 lang=csharp
 *
 * [39] Combination Sum
 */

// @lc code=start

using System;
using System.Collections.Generic;

public class Solution 
{
    private int[] candidates;

    private IList<IList<int>> results;

    public IList<IList<int>> CombinationSum(int[] candidates, int target) 
    {
        this.results = new List<IList<int>>();

        // sort candidates ascending
        this.candidates = candidates;
        Array.Sort(this.candidates);

        this.Combination(0, target, new Stack<int>());

        return this.results;
    }

    public void Combination(int start, int target, Stack<int> result)
    {
        // Console.WriteLine($"start: {start}, target: {target}");
        for (int i = start; i < this.candidates.Length; i++)
        {
            var candidate = this.candidates[i];
            var subTarget = target - candidate;
            if (subTarget < 0)
            {
                return;
            }
            else if (subTarget == 0)
            {
                result.Push(candidate);
                // each result in reverse order, doesn't matter in this case.
                this.results.Add(result.ToList()); 
                result.Pop();
            }
            else 
            {
                result.Push(candidate);
                this.Combination(i, subTarget, result);
                result.Pop();
            }
        }
    }
}
// @lc code=end

